### DISCUSSION TITLE
Database Naming Standards

### DISCUSSION THREAD

**Discussion by llong2195** [cortip/isomera#135] [General]
**Is your feature request related to a problem?** 
I want to introduce the rules for naming entities and columns
This helps everything look clearer
I can monitor the column without having to go into the database or migration file to see it

**Describe the solution you'd like**

     @Column({ name: 'user_id', type: 'int', unsigned: true, default: () => "'0'" })
     userId: number;`

     @Column({ name: 'user_name', type: 'varchar', nullable: true, length: 512 })
     userName: string;
We have to code and rewrite a lot more, but compared to the benefits it seems great :) 

**Describe alternatives you've considered** 
No ideas
**Additional context**

**Comment by vygandas** [comment]
@llong2195 great initiative :)
Let's figure this out: why do we need to mix camel and snake cases?
I mean, why this isn't ok 👇🏻 ?
```
 @Column({ name: 'userName', type: 'varchar', nullable: true, length: 512 })
 userName: string;
```

**Comment by llong2195** [comment]
@vygandas No problem, it seems I'm used to working with mix =))

**Comment by vygandas** [comment]
@llong2195 One more thing - since we do not use automatic sync and rely on migrations, we don't really need to put all these database column parameters. It won't generate types or won't do anything with the database itself. Unless I don't know something myself lol 😅

**Comment by llong2195** [comment]
@vygandas Just a few important parameters like `type`, `default`, `comment` or `name` if using `mix` like me

**Comment by llong2195** [comment]
@vygandas It's scary to use auto-sync on a production environment 🤣

**Comment by vygandas** [comment]
@llong2195 so many things could go wrong with auto 🤣 

![image](https://github.com/cortip/isomera/assets/723200/fea4a8de-a1d1-4a57-9dee-96417490ab04)

**Comment by llong2195** [comment]
@vygandas As I described, this just makes it easier for us to monitor the database

**Comment by llong2195** [comment]
@vygandas Considering the benefits it brings, should we use it?
We have to write more, edit more after each migration update...

**Comment by vygandas** [comment]
@llong2195 I've started digging a bit to find more info. Take a look at this one https://www.reddit.com/r/laravel/comments/dd8l90/using_snake_case_database_fields_is_making/ . It's for Laravel, but practically, they are discussing the same problem. It is also reflecting the same concern I have regarding different naming conventions and consistency.

**Comment by llong2195** [comment]
@vygandas Now both are fine.
So which option do you think is suitable for the project?

**Comment by vygandas** [comment]
@mantaskaveckas @algirdaspaskevicius @naftalimurgor @SpaceHolderdJs, what is your position on this?

**Comment by algirdaspaskevicius** [comment]
I think that a naming convention should definitely be agreed, however, which one is a question of preference.

Personally, I prefer `snake_case` naming strategy for database, the first Stack Overflow answer explains why:

https://stackoverflow.com/questions/2878248/postgresql-naming-conventions

Additionally, if I remember correctly, TypeORM will add some additional data like "entity" in the table name, which I personally don't like. I think that database should have a structure, that is acceptable in the SQL environment, because there may be cases where you may interact with the database without ORM.

Finally, there is no need to add column names manually, just use the SnakeNamingStrategy in the ORM config:

`import { SnakeNamingStrategy } from 'typeorm-naming-strategies';`

And in your ORM config, specify the naming strategy:

`const ormConfig = {
    ...otherConfigData,
   namingStrategy: new SnakeNamingStrategy()
}`

Note, that there are other naming strategies as well.

**Reply by vygandas** [reply]
Yep, I agree, manually naming each column seems like an overkill , especially when we do not rely or use auto sync from entities anyway. Need to keep as clean as possible.

**Reply by vygandas** [reply]
@llong2195 check this out. I guess you were right ;) Implementing a strategy sounds good. I just don't want you to add that many params in annotations. 😅  Should be as few as possible. If it's auto-generated, then there's no need for anything extra at all. What do you think?

**Reply by llong2195** [reply]
Thanks @vygandas  @algirdaspaskevicius , sounds good, I'll test it.

**Reply by llong2195** [reply]
@vygandas After some research, it seems that when we pre-define the name for the column, there is no need to go through snakeCase().
[typeorm-naming-strategies](https://github.com/tonivj5/typeorm-naming-strategies/blob/1a27ba8b9a5a98de87b9b8769c75d3031a6ddaa9/src/snake-naming.strategy.ts#L20), [typeorm](https://github.com/typeorm/typeorm/blob/d184d8598c057ce8fa54815e669b567238f3a86e/src/naming-strategy/DefaultNamingStrategy.ts#L44)
So by using the name: 'snake_case' the speed will look a bit faster

**Reply by vygandas** [reply]
@llong2195 I think that's overkill. We aim to make this as easy as possible for developers who will take this starter. I'd go with adding a naming strategy and keeping the code thinner.

**Reply by algirdaspaskevicius** [reply]
Well, defining the naming strategy in ORM config, will take care of the `snake_case` style in the database for the tables and columns. It will also take care of the `snake_case` strategy in the migrations. On the other hand, the code part will still have defined `camelCase` so the data from the database will still be returned as `someField` while in the database it will look like `some_field`. So simply defining the naming strategy will take care of everything.

**Reply by vygandas** [reply]
@algirdaspaskevicius , ok sounds nice. One question regarding mentioned migrations.. Will it convert this https://github.com/cortip/isomera/blob/main/apps/api/src/migrations/1700759687513-user-basic-table.ts#L27 or you'd need to write snake_case in that `name` field?

**Reply by vygandas** [reply]
@llong2195 btw, could you ping me on discord please?
